File: test/org/stringtemplate/v4/test/TestSubtemplates.java
/*
 [The "BSD license"]
 Copyright (c) 2009 Terence Parr
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.stringtemplate.v4.test;

import org.junit.*;
import org.stringtemplate.v4.ST;
import org.stringtemplate.v4.STGroup;
import org.stringtemplate.v4.STGroupFile;
import org.stringtemplate.v4.misc.ErrorBuffer;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;

public class TestSubtemplates extends BaseTest {

    @Test public void testSimpleIteration()
    {
        STGroup group = new STGroup();
        group.defineTemplate("test", "names", "<names:{n|<n>}>!");
        ST st = group.getInstanceOf("test");
        st.add("names", "Ter");
        st.add("names", "Tom");
        st.add("names", "Sumana");
        String expected = "TerTomSumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testMapIterationIsByKeys()
    {
        STGroup group = new STGroup();
        group.defineTemplate("test", "emails", "<emails:{n|<n>}>!");
        ST st = group.getInstanceOf("test");
        Map<String,String> emails = new LinkedHashMap<String,String>();
        emails.put("parrt", "Ter");
        emails.put("tombu", "Tom");
        emails.put("dmose", "Dan");
        st.add("emails", emails);
        String expected = "parrttombudmose!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testSimpleIterationWithArg()
    {
        STGroup group = new STGroup();
        group.defineTemplate("test", "names", "<names:{n | <n>}>!");
        ST st = group.getInstanceOf("test");
        st.add("names", "Ter");
        st.add("names", "Tom");
        st.add("names", "Sumana");
        String expected = "TerTomSumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testNestedIterationWithArg()
    {
        STGroup group = new STGroup();
        group.defineTemplate("test", "users", "<users:{u | <u.id:{id | <id>=}><u.name>}>!");
        ST st = group.getInstanceOf("test");
        st.add("users", new TestCoreBasics.User(1, "parrt"));
        st.add("users", new TestCoreBasics.User(2, "tombu"));
        st.add("users", new TestCoreBasics.User(3, "sri"));
        String expected = "1=parrt2=tombu3=sri!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testSubtemplateAsDefaultArg()
    {
        String templates =
            "t(x,y={<x:{s|<s><s>}>}) ::= <<\n" +
            "x: <x>\n" +
            "y: <y>\n" +
            ">>"+newline
            ;
        writeFile(tmpdir, "group.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/group.stg");
        ST b = group.getInstanceOf("t");
        b.add("x", "a");
        String expecting =
            "x: a" +newline+
            "y: aa";
        String result = b.render();
        assertEquals(expecting, result);
    }

    @Test public void testParallelAttributeIteration()
    {
        ST e = new ST(
                "<names,phones,salaries:{n,p,s | <n>@<p>: <s>\n}>"
            );
        e.add("names", "Ter");
        e.add("names", "Tom");
        e.add("phones", "1");
        e.add("phones", "2");
        e.add("salaries", "big");
        e.add("salaries", "huge");
        String expecting = "Ter@1: big"+newline+"Tom@2: huge"+newline;
        assertEquals(expecting, e.render());
    }

    @Test public void testParallelAttributeIterationWithNullValue()
    {
        ST e = new ST(
                "<names,phones,salaries:{n,p,s | <n>@<p>: <s>\n}>"
            );
        e.add("names", "Ter");
        e.add("names", "Tom");
        e.add("names", "Sriram");
        e.add("phones", new ArrayList<String>() {{add("1"); add(null); add("3");}});
        e.add("salaries", "big");
        e.add("salaries", "huge");
        e.add("salaries", "enormous");
        String expecting = "Ter@1: big"+newline+
                           "Tom@: huge"+newline+
                           "Sriram@3: enormous"+newline;
        assertEquals(expecting, e.render());
    }

    @Test public void testParallelAttributeIterationHasI()
    {
        ST e = new ST(
                "<names,phones,salaries:{n,p,s | <i0>. <n>@<p>: <s>\n}>"
            );
        e.add("names", "Ter");
        e.add("names", "Tom");
        e.add("phones", "1");
        e.add("phones", "2");
        e.add("salaries", "big");
        e.add("salaries", "huge");
        String expecting =
            "0. Ter@1: big"+newline+
            "1. Tom@2: huge"+newline;
        assertEquals(expecting, e.render());
    }

    @Test public void testParallelAttributeIterationWithDifferentSizes()
    {
        ST e = new ST(
                "<names,phones,salaries:{n,p,s | <n>@<p>: <s>}; separator=\", \">"
            );
        e.add("names", "Ter");
        e.add("names", "Tom");
        e.add("names", "Sriram");
        e.add("phones", "1");
        e.add("phones", "2");
        e.add("salaries", "big");
        String expecting = "Ter@1: big, Tom@2: , Sriram@: ";
        assertEquals(expecting, e.render());
    }

    @Test public void testParallelAttributeIterationWithSingletons()
    {
        ST e = new ST(
                "<names,phones,salaries:{n,p,s | <n>@<p>: <s>}; separator=\", \">"
            );
        e.add("names", "Ter");
        e.add("phones", "1");
        e.add("salaries", "big");
        String expecting = "Ter@1: big";
        assertEquals(expecting, e.render());
    }

    @Test public void testParallelAttributeIterationWithDifferentSizesTemplateRefInsideToo()
    {
        String templates =
                "page(names,phones,salaries) ::= "+newline+
                "   << <names,phones,salaries:{n,p,s | <value(n)>@<value(p)>: <value(s)>}; separator=\", \"> >>"+newline +
                "value(x) ::= \"<if(!x)>n/a<else><x><endif>\"" +newline;
        writeFile(tmpdir, "g.stg", templates);

        STGroup group = new STGroupFile(tmpdir+"/g.stg");
        ST p = group.getInstanceOf("page");
        p.add("names", "Ter");
        p.add("names", "Tom");
        p.add("names", "Sriram");
        p.add("phones", "1");
        p.add("phones", "2");
        p.add("salaries", "big");
        String expecting = " Ter@1: big, Tom@2: n/a, Sriram@n/a: n/a ";
        assertEquals(expecting, p.render());
    }

    @Test public void testEvalSTIteratingSubtemplateInSTFromAnotherGroup()
    {
        ErrorBuffer errors = new ErrorBuffer();
        STGroup innerGroup = new STGroup();
        innerGroup.setListener(errors);
        innerGroup.defineTemplate("test", "m", "<m:samegroup()>");
        innerGroup.defineTemplate("samegroup", "x", "hi ");
        ST st = innerGroup.getInstanceOf("test");
        st.add("m", new int[] {1,2,3});

        STGroup outerGroup = new STGroup();
        outerGroup.defineTemplate("errorMessage", "x", "<x>");
        ST outerST = outerGroup.getInstanceOf("errorMessage");
        outerST.add("x", st);

        String expected = "hi hi hi ";
        String result = outerST.render();

        assertEquals(errors.errors.size(), 0); // ignores no such prop errors

        assertEquals(expected, result);
    }

    @Test public void testEvalSTIteratingSubtemplateInSTFromAnotherGroupSingleValue()
    {
        ErrorBuffer errors = new ErrorBuffer();
        STGroup innerGroup = new STGroup();
        innerGroup.setListener(errors);
        innerGroup.defineTemplate("test", "m", "<m:samegroup()>");
        innerGroup.defineTemplate("samegroup", "x", "hi ");
        ST st = innerGroup.getInstanceOf("test");
        st.add("m", 10);

        STGroup outerGroup = new STGroup();
        outerGroup.defineTemplate("errorMessage", "x", "<x>");
        ST outerST = outerGroup.getInstanceOf("errorMessage");
        outerST.add("x", st);

        String expected = "hi ";
        String result = outerST.render();

        assertEquals(errors.errors.size(), 0); // ignores no such prop errors

        assertEquals(expected, result);
    }

    @Test public void testEvalSTFromAnotherGroup()
    {
        ErrorBuffer errors = new ErrorBuffer();
        STGroup innerGroup = new STGroup();
        innerGroup.setListener(errors);
        innerGroup.defineTemplate("bob", "inner");
        ST st = innerGroup.getInstanceOf("bob");

        STGroup outerGroup = new STGroup();
        outerGroup.setListener(errors);
        outerGroup.defineTemplate("errorMessage", "x", "<x>");
        outerGroup.defineTemplate("bob", "outer"); // should not be visible to test() in innerGroup
        ST outerST = outerGroup.getInstanceOf("errorMessage");
        outerST.add("x", st);

        String expected = "inner";
        String result = outerST.render();

        assertEquals(errors.errors.size(), 0); // ignores no such prop errors

        assertEquals(expected, result);
    }

}

File: objc/main/test/output/TestSubtemplates.m
#import "TestSubtemplates.h"

@implementation TestSubtemplates_Anon1

- (void) init {
  if (self = [super init]) {
    [self add:@"1"];
    [self add:nil];
    [self add:@"3"];
  }
  return self;
}

@end

@implementation TestSubtemplates

- (void) testSimpleIteration {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"names" param2:@"<names:{n|<n>}>!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"names" param1:@"Ter"];
  [st add:@"names" param1:@"Tom"];
  [st add:@"names" param1:@"Sumana"];
  NSString * expected = @"TerTomSumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testMapIterationIsByKeys {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"emails" param2:@"<emails:{n|<n>}>!"];
  ST * st = [group getInstanceOf:@"test"];
  NSMutableDictionary * emails = [[[LinkedHashMap alloc] init] autorelease];
  [emails setObject:@"parrt" param1:@"Ter"];
  [emails setObject:@"tombu" param1:@"Tom"];
  [emails setObject:@"dmose" param1:@"Dan"];
  [st add:@"emails" param1:emails];
  NSString * expected = @"parrttombudmose!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testSimpleIterationWithArg {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"names" param2:@"<names:{n | <n>}>!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"names" param1:@"Ter"];
  [st add:@"names" param1:@"Tom"];
  [st add:@"names" param1:@"Sumana"];
  NSString * expected = @"TerTomSumana!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testNestedIterationWithArg {
  STGroup * group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"test" param1:@"users" param2:@"<users:{u | <u.id:{id | <id>=}><u.name>}>!"];
  ST * st = [group getInstanceOf:@"test"];
  [st add:@"users" param1:[[[User alloc] init:1 param1:@"parrt"] autorelease]];
  [st add:@"users" param1:[[[User alloc] init:2 param1:@"tombu"] autorelease]];
  [st add:@"users" param1:[[[User alloc] init:3 param1:@"sri"] autorelease]];
  NSString * expected = @"1=parrt2=tombu3=sri!";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) testSubtemplateAsDefaultArg {
  NSString * templates = [[[@"t(x,y={<x:{s|<s><s>}>}) ::= <<\n" stringByAppendingString:@"x: <x>\n"] stringByAppendingString:@"y: <y>\n"] stringByAppendingString:@">>"] + newline;
  [self writeFile:tmpdir param1:@"group.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[tmpdir stringByAppendingString:@"/group.stg"]] autorelease];
  ST * b = [group getInstanceOf:@"t"];
  [b add:@"x" param1:@"a"];
  NSString * expecting = [[@"x: a" stringByAppendingString:newline] stringByAppendingString:@"y: aa"];
  NSString * result = [b render];
  [self assertEquals:expecting param1:result];
}

- (void) testParallelAttributeIteration {
  ST * e = [[[ST alloc] init:@"<names,phones,salaries:{n,p,s | <n>@<p>: <s>\n}>"] autorelease];
  [e add:@"names" param1:@"Ter"];
  [e add:@"names" param1:@"Tom"];
  [e add:@"phones" param1:@"1"];
  [e add:@"phones" param1:@"2"];
  [e add:@"salaries" param1:@"big"];
  [e add:@"salaries" param1:@"huge"];
  NSString * expecting = [[@"Ter@1: big" stringByAppendingString:newline] stringByAppendingString:@"Tom@2: huge"] + newline;
  [self assertEquals:expecting param1:[e render]];
}

- (void) testParallelAttributeIterationWithNullValue {
  ST * e = [[[ST alloc] init:@"<names,phones,salaries:{n,p,s | <n>@<p>: <s>\n}>"] autorelease];
  [e add:@"names" param1:@"Ter"];
  [e add:@"names" param1:@"Tom"];
  [e add:@"names" param1:@"Sriram"];
  [e add:@"phones" param1:[[[TestSubtemplates_Anon1 alloc] init] autorelease]];
  [e add:@"salaries" param1:@"big"];
  [e add:@"salaries" param1:@"huge"];
  [e add:@"salaries" param1:@"enormous"];
  NSString * expecting = [[[@"Ter@1: big" stringByAppendingString:newline] stringByAppendingString:@"Tom@: huge"] + newline stringByAppendingString:@"Sriram@3: enormous"] + newline;
  [self assertEquals:expecting param1:[e render]];
}

- (void) testParallelAttributeIterationHasI {
  ST * e = [[[ST alloc] init:@"<names,phones,salaries:{n,p,s | <i0>. <n>@<p>: <s>\n}>"] autorelease];
  [e add:@"names" param1:@"Ter"];
  [e add:@"names" param1:@"Tom"];
  [e add:@"phones" param1:@"1"];
  [e add:@"phones" param1:@"2"];
  [e add:@"salaries" param1:@"big"];
  [e add:@"salaries" param1:@"huge"];
  NSString * expecting = [[@"0. Ter@1: big" stringByAppendingString:newline] stringByAppendingString:@"1. Tom@2: huge"] + newline;
  [self assertEquals:expecting param1:[e render]];
}

- (void) testParallelAttributeIterationWithDifferentSizes {
  ST * e = [[[ST alloc] init:@"<names,phones,salaries:{n,p,s | <n>@<p>: <s>}; separator=\", \">"] autorelease];
  [e add:@"names" param1:@"Ter"];
  [e add:@"names" param1:@"Tom"];
  [e add:@"names" param1:@"Sriram"];
  [e add:@"phones" param1:@"1"];
  [e add:@"phones" param1:@"2"];
  [e add:@"salaries" param1:@"big"];
  NSString * expecting = @"Ter@1: big, Tom@2: , Sriram@: ";
  [self assertEquals:expecting param1:[e render]];
}

- (void) testParallelAttributeIterationWithSingletons {
  ST * e = [[[ST alloc] init:@"<names,phones,salaries:{n,p,s | <n>@<p>: <s>}; separator=\", \">"] autorelease];
  [e add:@"names" param1:@"Ter"];
  [e add:@"phones" param1:@"1"];
  [e add:@"salaries" param1:@"big"];
  NSString * expecting = @"Ter@1: big";
  [self assertEquals:expecting param1:[e render]];
}

- (void) testParallelAttributeIterationWithDifferentSizesTemplateRefInsideToo {
  NSString * templates = [[[@"page(names,phones,salaries) ::= " stringByAppendingString:newline] stringByAppendingString:@"	<< <names,phones,salaries:{n,p,s | <value(n)>@<value(p)>: <value(s)>}; separator=\", \"> >>"] + newline stringByAppendingString:@"value(x) ::= \"<if(!x)>n/a<else><x><endif>\""] + newline;
  [self writeFile:tmpdir param1:@"g.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[tmpdir stringByAppendingString:@"/g.stg"]] autorelease];
  ST * p = [group getInstanceOf:@"page"];
  [p add:@"names" param1:@"Ter"];
  [p add:@"names" param1:@"Tom"];
  [p add:@"names" param1:@"Sriram"];
  [p add:@"phones" param1:@"1"];
  [p add:@"phones" param1:@"2"];
  [p add:@"salaries" param1:@"big"];
  NSString * expecting = @" Ter@1: big, Tom@2: n/a, Sriram@n/a: n/a ";
  [self assertEquals:expecting param1:[p render]];
}

- (void) testEvalSTIteratingSubtemplateInSTFromAnotherGroup {
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  STGroup * innerGroup = [[[STGroup alloc] init] autorelease];
  [innerGroup setListener:errors];
  [innerGroup defineTemplate:@"test" param1:@"m" param2:@"<m:samegroup()>"];
  [innerGroup defineTemplate:@"samegroup" param1:@"x" param2:@"hi "];
  ST * st = [innerGroup getInstanceOf:@"test"];
  [st add:@"m" param1:[NSArray arrayWithObjects:1, 2, 3, nil]];
  STGroup * outerGroup = [[[STGroup alloc] init] autorelease];
  [outerGroup defineTemplate:@"errorMessage" param1:@"x" param2:@"<x>"];
  ST * outerST = [outerGroup getInstanceOf:@"errorMessage"];
  [outerST add:@"x" param1:st];
  NSString * expected = @"hi hi hi ";
  NSString * result = [outerST render];
  [self assertEquals:[errors.errors size] param1:0];
  [self assertEquals:expected param1:result];
}

- (void) testEvalSTIteratingSubtemplateInSTFromAnotherGroupSingleValue {
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  STGroup * innerGroup = [[[STGroup alloc] init] autorelease];
  [innerGroup setListener:errors];
  [innerGroup defineTemplate:@"test" param1:@"m" param2:@"<m:samegroup()>"];
  [innerGroup defineTemplate:@"samegroup" param1:@"x" param2:@"hi "];
  ST * st = [innerGroup getInstanceOf:@"test"];
  [st add:@"m" param1:10];
  STGroup * outerGroup = [[[STGroup alloc] init] autorelease];
  [outerGroup defineTemplate:@"errorMessage" param1:@"x" param2:@"<x>"];
  ST * outerST = [outerGroup getInstanceOf:@"errorMessage"];
  [outerST add:@"x" param1:st];
  NSString * expected = @"hi ";
  NSString * result = [outerST render];
  [self assertEquals:[errors.errors size] param1:0];
  [self assertEquals:expected param1:result];
}

- (void) testEvalSTFromAnotherGroup {
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  STGroup * innerGroup = [[[STGroup alloc] init] autorelease];
  [innerGroup setListener:errors];
  [innerGroup defineTemplate:@"bob" param1:@"inner"];
  ST * st = [innerGroup getInstanceOf:@"bob"];
  STGroup * outerGroup = [[[STGroup alloc] init] autorelease];
  [outerGroup setListener:errors];
  [outerGroup defineTemplate:@"errorMessage" param1:@"x" param2:@"<x>"];
  [outerGroup defineTemplate:@"bob" param1:@"outer"];
  ST * outerST = [outerGroup getInstanceOf:@"errorMessage"];
  [outerST add:@"x" param1:st];
  NSString * expected = @"inner";
  NSString * result = [outerST render];
  [self assertEquals:[errors.errors size] param1:0];
  [self assertEquals:expected param1:result];
}

@end
